kafka-connect-aws-s3/src/main/scala/io/lenses/streamreactor/connect/aws/s3/formats/reader/SourceData.scala

```diff
@@ -35,31 +36,40 @@ abstract class SourceData(lineNumber: Long) {
   def getLineNumber: Long = lineNumber
 
   def toSourceRecord(
-    bucketAndPath: RemoteS3PathLocation,
+    bucketAndPath: S3Location,
     targetTopic:   String,
     partitionFn:   String => Option[Int],
-  ): SourceRecord = representationKey match {
-    case Some(key) =>
-      new SourceRecord(
-        fromSourcePartition(bucketAndPath.root()).asJava,
-        fromSourceOffset(bucketAndPath, getLineNumber).asJava,
-        targetTopic,
-        partitionFn(bucketAndPath.path).map(Int.box).orNull,
-        null,
-        key,
-        representationSchema.orNull,
-        representationValue,
-      )
-    case None =>
-      new SourceRecord(
-        fromSourcePartition(bucketAndPath.root()).asJava,
-        fromSourceOffset(bucketAndPath, getLineNumber).asJava,
-        targetTopic,
-        partitionFn(bucketAndPath.path).map(Int.box).orNull,
-        representationSchema.orNull,
-        representationValue,
-      )
-  }
+  ): Either[Throwable, SourceRecord] =
+    bucketAndPath
+      .path
+      .toRight(new IllegalArgumentException("No path exists in S3Location"))
+      .map {
+        pth =>
+          val partition = partitionFn(pth).map(Int.box).orNull
```

Comment:
The partition target is a function of the key as you know. In that case the partitionFN should use murmur3 (like Kafkadoes). But why not let Kafka producer take the work? If some stream processing happens downstream, it might be better.